src/shared/vscode/commands2.ts

```diff
@@ -0,0 +1,330 @@
+    readonly id: string
+    readonly name?: string
+    readonly factory: CommandFactory<T, U>
+}
+
+class CommandResource<T extends Callback = Callback, U extends any[] = any[]> {
+    private disposed?: boolean
+    private subscription?: vscode.Disposable
+    private static counter = 0 // Used to generated unique-ish tree item IDs
+    public readonly id = this.resource.id
+
+    public constructor(private readonly resource: Deferred<T, U>, private readonly commands = vscode.commands) {}
+
+    public get declared() {
+        return !this.disposed
+    }
+
+    public get registered() {
+        return !!this.subscription
+    }
+
+    public build(...args: Parameters<T>): Builder {
+        const id = this.resource.id
+
+        return {
+            asUri: this.buildUri(id, args),
+            asCommand: this.buildCommand(id, args),
+            asCodeLens: this.buildCodeLens(id, args),
+            asTreeNode: this.buildTreeNode(id, args),
+        }
+    }
+
+    public register(...args: U): RegisteredCommand<T> {
+        const { id, name } = this.resource
+        const label = name ? `"${name}" (id: ${id})` : `"${id}"`
+        const target = this.resource.factory(...args)
+        const instrumented = (...args: Parameters<T>) => runCommand(target, { label, args })
+        this.subscription = this.commands.registerCommand(this.resource.id, instrumented)
+
+        return this
+    }
+
+    public async execute(...args: Parameters<T>): Promise<ReturnType<T> | undefined> {
+        return this.commands.executeCommand<ReturnType<T>>(this.resource.id, ...args)
+    }
+
+    public dispose(): void {
+        this.disposed = true
+        this.subscription?.dispose()
+    }
+
+    private buildUri(id: string, args: unknown[]) {
+        return () => vscode.Uri.parse(`command:${id}?${encodeURIComponent(JSON.stringify(args))}`)
+    }
+
+    private buildCommand(id: string, args: unknown[]) {
+        return (content: PartialCommand) => ({ ...content, command: id, arguments: args })
+    }
+
+    private buildCodeLens(id: string, args: unknown[]) {
+        return (range: vscode.Range, content: PartialCommand) => {
+            return new vscode.CodeLens(range, this.buildCommand(id, args)(content))
+        }
+    }
+
+    private buildTreeNode(id: string, args: unknown[]) {
+        return (content: PartialTreeItem) => {
+            return new (class extends AWSTreeNodeBase {
+                public constructor() {
+                    super(content.label, vscode.TreeItemCollapsibleState.None)
+                    this.id = `${id}-${(CommandResource.counter += 1)}`
+                    this.command = { command: id, arguments: args, title: content.label }
+
+                    Object.assign(this, content)
+                }
+            })()
+        }
+    }
+}
+
+interface CommandInfo<T extends Callback> {
+    readonly label: string
+    readonly args: Parameters<T>
+    readonly logging?: boolean
```

Comment:
also `telemetry` ?

Reply:
It'll be another PR due to the additional complexity. We need to think of telemetry more of like a logger for it to work nicely in a cross-cutting fashion. The way we do it now means only commands without metadata would be easily supported. Which is maybe fine?

Reply:
sure the `telemetry` field would just be stub for now. 

> The way we do it now means only commands without metadata would be easily supported.

I was thinking we could have a generic metric like `userAction { name = ... }`. This could be opt-out for commands that want/need richer telemetry. But it probably covers a lot of actions that we are currently missing.

Reply:
Yeah seems reasonable. Majority of commands don't need extra info.